src/vos/evtree.c

```diff
@@ -1930,13 +1924,24 @@ evt_ent_array_fill(struct evt_context *tcx, enum evt_find_opc find_opc,
 			}
 			V_TRACE(DB_TRACE, "Found overlapped leaf rect\n");
 
+			desc = evt_node_desc_at(tcx, node, i);
+			rc = evt_dtx_check_visibility(tcx, desc, intent);
+			/* Skip the invisible record. */
+			if (rc == DTX_VBT_INVISIBLE)
+				continue;
+
 			/* early check */
 			switch (find_opc) {
 			default:
 				D_ASSERTF(0, "%d\n", find_opc);
 			case EVT_FIND_OVERWRITE:
 				if (time_overlap != RT_OVERLAP_SAME)
 					continue; /* not same epoch, skip */
+
+				/* If the visibility is unknown, then go out. */
+				if (rc < 0)
+					D_GOTO(out, rc);
+
```

Comment:
Similar to btree, I think the special 'visibility' check only required for EVT_FIND_OVERWRITE case (for clearing existing aborted leaf node).
It's actually not visibility check, it to check if an aborted leaf node can be cleared.

Reply:
> I agree with Jeff that the visibility check should be done by vos tree (when get leaf on fetch, iteration or preparing next level tree root) but not common btree.

> To support 'reusing' aborted record, we may introduce another interface to check if the leaf node is an aborted one which could be cleared, and this interface should only be called on CMP_EQ case, so that once eligible aborted record is detected on insert, we can clear it immediately. I suppose that requires less changes in common btree, also it won't expose the concept of 'visibility' to common btree (which looks kind of layer violation to me).


Move the visibility check into VOS is possible, but I do not think it makes things to be more clean, because the visibility check point is quite limited inside the btree.c, that is relative easy to control. If we move it into VOS, then we will have to check it everywhere when call probe/fetch/delete/iterate (forward and/or backward). It is quite possible that new callers may miss to check the visibility and cause trouble. On the other hand, if we have to add new btree API for checking aborted record for reusing when insert, then what is the main difference with current visibility API?


> Similar to btree, I think the special 'visibility' check only required for EVT_FIND_OVERWRITE case (for clearing existing aborted leaf node). It's actually not visibility check, it to check if an aborted leaf node can be cleared.

The visibility check for EVTree is not only for EVT_FIND_OVERWRITE case but also required for others, such as when compute MBR, inserting new record, and so on.